dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/maintenance/jdbc/JdbcMaintenanceStore.java

```diff
@@ -216,25 +251,55 @@ public int deleteSoftDeletedTrackedEntityInstances()
 
         if ( result > 0 && deletedTeiUids.size() > 0 )
         {
-            auditHardDeletedTrackedEntityInstances( deletedTeiUids );
+            auditHardDeletedEntity( deletedTeiUids, TrackedEntityInstance.class );
+
+            if ( associatedEnrollments.size() > 0 )
+            {
+                auditHardDeletedEntity( associatedEnrollments, ProgramInstance.class );
+            }
+
+            if ( associatedEvents.size() > 0 )
+            {
+                auditHardDeletedEntity( associatedEvents, ProgramStageInstance.class );
+            }
         }
 
         return result;
     }
 
-    private void auditHardDeletedTrackedEntityInstances( List<String> deletedTeiUids )
+    private List<String> getDeletionEntities( String entitySql )
+    {
+        /*
+         * Get all soft deleted entities before they are hard deleted from
+         * database
+         */
+        List<String> deletedUids = new ArrayList<>();
+
+        SqlRowSet softDeletedEntitiesUidRows = jdbcTemplate.queryForRowSet( entitySql );
+
+        while ( softDeletedEntitiesUidRows.next() )
+        {
+            deletedUids.add( softDeletedEntitiesUidRows.getString( "uid" ) );
+        }
+
+        return deletedUids;
+    }
+
+    private void auditHardDeletedEntity( List<String> deletedEnrollments, Class<? extends SoftDeletableObject> entity )
```

Comment:
The name of the argument should be generic based on its usage. deletedEntities or deletedUids or deletedEntityUids. Instead of deletedEnrollments